utils/frame/try-runtime/cli/src/lib.rs

```diff
@@ -0,0 +1,149 @@
+#[derive(Debug, structopt::StructOpt)]
+pub struct TryRuntimeCmd {
+	/// The shared parameters
+	#[allow(missing_docs)]
+	#[structopt(flatten)]
+	pub shared_params: sc_cli::SharedParams,
+
+	/// The target pallet to run the migration against.
+	#[structopt(short, long, default_value = "All")]
+	pub target: Target,
+
+	/// The state to use to run the migration. Should be a valid FILE or HTTP URI.
+	#[structopt(short, long, default_value = "http://localhost:9933")]
+	pub state: State,
+}
+
+/// The state to use for a migration dry-run.
+#[derive(Debug)]
+pub enum State {
+	/// A snapshot. Inner value is a file path.
+	Snap(String),
+
+	/// A live chain. Inner value is the HTTP uri.
+	Live(String),
+}
+
+impl FromStr for State {
+	type Err = &'static str;
+	fn from_str(s: &str) -> Result<Self, Self::Err> {
+		match s.get(..7) {
+			// could use Url crate as well, but lets keep it simple for now.
+			Some("http://") => Ok(State::Live(s.to_string())),
+			Some("file://") => s
+				.split("//")
+				.collect::<Vec<_>>()
+				.get(1)
+				.map(|s| State::Snap(s.to_string()))
+				.ok_or("invalid file URI"),
+			_ => Err("invalid format. Must be a valid HTTP or File URI"),
+		}
+	}
+}
+
+impl TryRuntimeCmd {
+	pub async fn run<B, ExecDispatch>(&self, config: Configuration) -> sc_cli::Result<()>
+	where
+		B: BlockT,
+		ExecDispatch: NativeExecutionDispatch + 'static,
+	{
+
+		let spec = config.chain_spec;
+		let genesis_storage = spec.build_storage()?;
+
+		let code = StorageData(genesis_storage.top.get(well_known_keys::CODE).unwrap().to_vec());
+		let code_key = StorageKey(well_known_keys::CODE.to_vec());
+
+		let wasm_method = WasmExecutionMethod::Interpreted;
+		let strategy = ExecutionStrategy::Native;
+		let mut changes = Default::default();
+
+		let heap_pages = Some(1024);
+		let executor = NativeExecutor::<ExecDispatch>::new(wasm_method, heap_pages, 2);
+
+		let ext = {
+			use remote_externalities::{Builder, Mode, CacheConfig, OfflineConfig, OnlineConfig};
+			let builder = match &self.state {
+				State::Snap(file_path) => Builder::new().mode(Mode::Offline(OfflineConfig {
+					cache: CacheConfig { name: file_path.into(), ..Default::default() },
+				})),
+				State::Live(http_uri) => Builder::new().mode(Mode::Online(OnlineConfig {
+					uri: http_uri.into(),
+					..Default::default()
+				})),
+			};
+
+			builder.inject(&[(code_key, code)]).build().await
+		};
+
+		let consumed_weight = StateMachine::<_, _, NumberFor<B>, _>::new(
+			&ext.backend,
+			None,
+			&mut changes,
+			&executor,
+			"TryRuntime_on_runtime_upgrade",
+			&self.target.encode(),
+			ext.extensions,
+			&sp_state_machine::backend::BackendRuntimeCode::new(&ext.backend)
+				.runtime_code()
+				.unwrap(),
+			sp_core::testing::TaskExecutor::new(),
+		)
+		.execute(strategy.into())
+		.unwrap();
+
+		let weight = <u64 as Decode>::decode(&mut &*consumed_weight).unwrap();
+		log::info!(
+			"try-runtime executed without errors. Consumed weight = {}",
```

Comment:
Is consumed weight alone very helpful?

It should be compared as some percent of the maximum weight for a block.

Reply:
sounds like a good improvement.